File: mobile/test/ui/features/connection/connection_view_model_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:stream_channel/stream_channel.dart';
import 'package:wheeldeck/data/repositories/connection_repository.dart';
import 'package:wheeldeck/data/repositories/server_discovery_repository.dart';
import 'package:wheeldeck/data/repositories/session_repository.dart';
import 'package:wheeldeck/domain/models/connection_mode.dart';
import 'package:wheeldeck/domain/models/connection_status.dart';
import 'package:wheeldeck/domain/models/connection_target.dart';
import 'package:wheeldeck/domain/models/discovered_server.dart';
import 'package:wheeldeck/domain/models/pairing_method.dart';
import 'package:wheeldeck/data/services/discovery.dart';
import 'package:wheeldeck/data/services/pairing.dart';
import 'package:wheeldeck/data/services/wheeldeck_client.dart';
import 'package:wheeldeck/ui/features/connection/view_models/connection_view_model.dart';

class _MemoryStore implements SessionTokenStore {
  String? token;

  @override
  Future<String?> load() async => token;

  @override
  Future<void> save(String value) async => token = value;
}

ConnectionViewModel buildViewModel({
  Future<StreamChannel<dynamic>> Function(Uri uri)? connect,
  ServerResolver? resolver,
  SessionTokenStore? store,
}) {
  final client = WheelDeckClient(deviceId: 'phone-1', connect: connect);
  final discovery = ServerDiscovery(resolve: resolver);
  final pairing = PairingController(
    store: store ?? _MemoryStore(),
    client: client,
  );
  return ConnectionViewModel(
    discoveryRepository: ServerDiscoveryRepository(discovery: discovery),
    connectionRepository: ConnectionRepository(client: client),
    sessionRepository: SessionRepository(pairing: pairing),
  );
}

void main() {
  group('ConnectionViewModel', () {
    test('refreshDiscovery updates the server list and notifies', () async {
      var notifyCount = 0;
      final viewModel = buildViewModel(
        resolver: () async => const [
          DiscoveredServer(host: '10.0.0.1', port: 8765, name: 'Desktop'),
        ],
      )..addListener(() => notifyCount++);

      await viewModel.refreshDiscovery();

      expect(viewModel.servers, hasLength(1));
      expect(viewModel.servers.first.name, 'Desktop');
      expect(notifyCount, greaterThan(0));
      viewModel.dispose();
    });

    test('connect without a session asks for a PIN', () async {
      final channel = StreamChannelController<dynamic>(sync: true);
      channel.foreign.stream.listen((_) {});
      final viewModel = buildViewModel(
        connect: (uri) async => channel.local,
      );

      await viewModel.connect(
        const ConnectionTarget(
          mode: ConnectionMode.manual,
          ipAddress: '10.0.0.1',
        ),
      );

      expect(viewModel.pairingChallenge, isNotNull);
      expect(viewModel.pairingChallenge!.method, PairingMethod.pin);

      await viewModel.disconnect();
      viewModel.dispose();
    });

    test('pairingError is set after a rejected PIN submission', () async {
      final channel = StreamChannelController<dynamic>(sync: true);
      channel.foreign.stream.listen((_) {});
      final viewModel = buildViewModel(
        connect: (uri) async => channel.local,
      );

      await viewModel.connect(
        const ConnectionTarget(
          mode: ConnectionMode.manual,
          ipAddress: '10.0.0.1',
        ),
      );
      channel.foreign.sink.add(
        jsonEncode({'type': 'pair_response', 'accepted': false}),
      );

      viewModel.submitPairingCode('123456');
      expect(viewModel.pairingError, isFalse);

      channel.foreign.sink.add(
        jsonEncode({'type': 'pair_response', 'accepted': false}),
      );
      expect(viewModel.pairingError, isTrue);

      viewModel.submitPairingCode('999999');
      expect(viewModel.pairingError, isFalse);

      await viewModel.disconnect();
      viewModel.dispose();
    });

    test('pairing challenge is cleared on successful connection', () async {
      final channel = StreamChannelController<dynamic>(sync: true);
      channel.foreign.stream.listen((_) {});
      final viewModel = buildViewModel(
        connect: (uri) async => channel.local,
      );

      await viewModel.connect(
        const ConnectionTarget(
          mode: ConnectionMode.manual,
          ipAddress: '10.0.0.1',
        ),
      );
      channel.foreign.sink.add(
        jsonEncode({'type': 'pair_response', 'accepted': false}),
      );
      expect(viewModel.pairingChallenge, isNotNull);

      viewModel.submitPairingCode('123456');
      channel.foreign.sink.add(jsonEncode({
        'type': 'pair_response',
        'accepted': true,
        'session_token': 'tok-1',
      }));

      expect(viewModel.pairingChallenge, isNull);
      expect(viewModel.status, ConnectionStatus.connected);

      await viewModel.disconnect();
      viewModel.dispose();
    });
  });
}

File: android/app/src/test/kotlin/dev/fazrigading/wheeldeck/ui/features/connection/view_models/ConnectionViewModelTest.kt
package dev.fazrigading.wheeldeck.ui.features.connection.view_models

import dev.fazrigading.wheeldeck.data.repositories.ConnectionRepository
import dev.fazrigading.wheeldeck.data.repositories.PairedDeviceRepository
import dev.fazrigading.wheeldeck.data.repositories.ServerDiscoveryRepository
import dev.fazrigading.wheeldeck.data.repositories.SessionRepository
import dev.fazrigading.wheeldeck.data.services.PairingController
import dev.fazrigading.wheeldeck.data.services.ServerDiscovery
import dev.fazrigading.wheeldeck.data.services.ServerResolver
import dev.fazrigading.wheeldeck.data.services.SessionTokenStore
import dev.fazrigading.wheeldeck.data.services.WheelDeckClient
import dev.fazrigading.wheeldeck.domain.models.ConnectionMode
import dev.fazrigading.wheeldeck.domain.models.ConnectionStatus
import dev.fazrigading.wheeldeck.domain.models.ConnectionTarget
import dev.fazrigading.wheeldeck.domain.models.DiscoveredServer
import dev.fazrigading.wheeldeck.domain.models.PairingChallenge
import dev.fazrigading.wheeldeck.domain.models.PairingMethod
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.test.StandardTestDispatcher
import kotlinx.coroutines.test.advanceUntilIdle
import kotlinx.coroutines.test.resetMain
import kotlinx.coroutines.test.runTest
import kotlinx.coroutines.test.setMain
import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test

@OptIn(ExperimentalCoroutinesApi::class)
class ConnectionViewModelTest {
    private val dispatcher = StandardTestDispatcher()

    @Before
    fun setUp() {
        Dispatchers.setMain(dispatcher)
    }

    @After
    fun tearDown() {
        Dispatchers.resetMain()
    }

    private class FakeResolver(private val servers: List<DiscoveredServer> = emptyList()) : ServerResolver {
        override suspend fun resolve(): List<DiscoveredServer> = servers
    }

    private class FakeTokenStore : SessionTokenStore {
        var saved: String? = null
        override suspend fun load(): String? = null
        override suspend fun save(token: String) {
            saved = token
        }
        override suspend fun clear() {
            saved = null
        }
    }

    private class FakePairedStore : dev.fazrigading.wheeldeck.data.repositories.PairedDeviceStore {
        val paired = mutableSetOf<String>()
        override suspend fun load(): Set<String> = paired
        override suspend fun addPaired(host: String, port: Int) {
            paired.add("$host:$port")
        }
        override suspend fun removePaired(host: String, port: Int) {
            paired.remove("$host:$port")
        }
    }

    private class Fixture(servers: List<DiscoveredServer> = emptyList()) {
        val dispatcher = StandardTestDispatcher()
        val scope = CoroutineScope(dispatcher)
        val client = WheelDeckClient(deviceId = "test-device", scope = scope)
        val tokenStore = FakeTokenStore()
        val pairedStore = FakePairedStore()
        val vm = ConnectionViewModel(
            discoveryRepository = ServerDiscoveryRepository(ServerDiscovery(FakeResolver(servers))),
            connectionRepository = ConnectionRepository(client),
            sessionRepository = SessionRepository(PairingController(tokenStore, client, scope)),
            pairedDeviceRepository = PairedDeviceRepository(pairedStore),
        )
    }

    @Test
    fun `initial status comes from client`() = runTest(dispatcher) {
        val f = Fixture()
        assertEquals(ConnectionStatus.Disconnected, f.vm.uiState.value.status)
    }

    @Test
    fun `connect transitions status through Connecting`() = runTest(dispatcher) {
        val f = Fixture()
        f.vm.connect(ConnectionTarget(mode = ConnectionMode.Manual, ipAddress = "127.0.0.1", port = 1))
        advanceUntilIdle()
        assertEquals(ConnectionStatus.Connecting, f.vm.uiState.value.status)
    }

    @Test
    fun `refreshDiscovery updates servers`() = runTest(dispatcher) {
        val f = Fixture(listOf(DiscoveredServer(host = "192.168.1.10", port = 8765, name = "desktop")))

        f.vm.refreshDiscovery()
        advanceUntilIdle()

        assertEquals(1, f.vm.uiState.value.servers.size)
        assertTrue(f.vm.uiState.value.unpairedServers.isNotEmpty())
        assertTrue(f.vm.uiState.value.pairedServers.isEmpty())
    }

    @Test
    fun `onPairingRequired sets challenge`() = runTest(dispatcher) {
        val f = Fixture()
        assertNull(f.vm.uiState.value.pairingChallenge)

        f.vm.onPairingRequired(PairingChallenge(PairingMethod.Pin))
        assertEquals(PairingMethod.Pin, f.vm.uiState.value.pairingChallenge?.method)
    }

    @Test
    fun `submitPairingCode clears pairingError`() = runTest(dispatcher) {
        val f = Fixture()
        f.vm.onPairingRequired(PairingChallenge(PairingMethod.Pin))

        f.vm.submitPairingCode("1234")
        advanceUntilIdle()

        assertFalse(f.vm.uiState.value.pairingError)
    }

    @Test
    fun `pause and resume toggle isPaused`() = runTest(dispatcher) {
        val f = Fixture()
        assertFalse(f.vm.uiState.value.isPaused)

        f.vm.pause()
        assertTrue(f.vm.uiState.value.isPaused)

        f.vm.resume()
        assertFalse(f.vm.uiState.value.isPaused)
    }

    @Test
    fun `disconnect returns to Disconnected`() = runTest(dispatcher) {
        val f = Fixture()
        f.vm.disconnect()
        advanceUntilIdle()
        assertEquals(ConnectionStatus.Disconnected, f.vm.uiState.value.status)
    }
}
